ubuntu_image/parser.py

```diff
@@ -320,41 +332,56 @@ def parse(stream_or_string):
             offset_write = structure.get('offset-write')
             size = structure['size']
             structure_type = structure['type']
-            # Validate structure types.  These can be either GUIDs, two hex
-            # digits, hybrids, or the special 'mbr' type.  The basic syntactic
-            # validation happens above in the Voluptuous schema, but here we
-            # need to ensure cross-attribute constraints.  Specifically,
-            # hybrids and 'mbr' are allowed for either schema, but GUID-only
-            # is only allowed for GPT, while 2-digit-only is only allowed for
-            # MBR.  Note too that 2-item tuples are also already ensured.
+            structure_role = structure.get('role')
+            # Validate structure types and roles. These can be either GUIDs,
+            # two hex digits or hybrids.  The basic syntactic validation
+            # happens above in the Voluptuous schema, but here we need to
+            # ensure cross-attribute constraints.  Specifically, hybrids are
+            # allowed for either schema, but GUID-only is only allowed for GPT,
+            # while 2-digit-only is only allowed for MBR roles.
+            # Note too that 2-item tuples are also already ensured.
+            # We also still support the deprecated special type 'mbr' for
+            # legacy purposes, but issue a warning
+            if structure_type == 'mbr':
+                warn("volumes:<volume name>:structure:<N>:type = 'mbr' is "
+                     'deprecated; use role instead', DeprecationWarning)
+                if structure_role:
```

Comment:
Maybe we should move this to above the warning?  The way it is now, if you define both role and type, you'll get the deprecation warning and then the error.  I think it would be better to just show the error (since they're both encouraging the user to switch to role).